# FastGxC: A powerful and computationally efficient software for context-specific eQTL mapping in single-cell genomics data

FastGxC is a computationally efficient R package for mapping context-specific eQTLs across multiple biological conditions (e.g., tissues or cell types). It was originally designed for single-cell RNA-seq data, but also supports bulk RNA-seq from multi-context samples.

# Installation and Setup

Before using FastGxC, make sure all required R packages are installed.

To install FastGxC:

```         
devtools::install_github("BalliuLab/FastGxC")
library(FastGxC)
```

If you need help installing dependencies or Python integration, refer to the Wiki's [Installation and Dependences Guide](https://github.com/BalliuLab/FastGxC/wiki/1.-Installation-and-Dependencies)

# Installation and Setup

This repository is supported by a full [Wiki](https://github.com/BalliuLab/FastGxC/wiki) that includes the following:

1.  [Installation and Dependences](https://github.com/BalliuLab/FastGxC/wiki/1.-Installation-and-Dependencies)

2.  [Simulate Data](https://github.com/BalliuLab/FastGxC/wiki/2.-Simulate-Data)

3.  [Decomposition Step](https://github.com/BalliuLab/FastGxC/wiki/3.-Decomposition-Step)

4.  [eQTL Mapping](https://github.com/BalliuLab/FastGxC/wiki/4.-eQTL-Mapping)

5.  [Multiple Testing Correction](https://github.com/BalliuLab/FastGxC/wiki/5.-Multiple-Testing-Correction)

If you're unsure where to start or encounter issues, please check the [Wiki](https://github.com/BalliuLab/FastGxC/wiki) first â€” it covers all major steps with examples and explanations.
