# FastGxC: A powerful and computationally efficient software for context-specific eQTL mapping in single-cell omics data 

FastGxC was originally developed for single-cell data, where each individual contributes gene expression measurements across multiple cell types. 

However, it can also be applied to bulk RNA-seq data when the same individuals are profiled across multiple tissues or conditions. 

In both settings, FastGxC models **repeated samples** from each individual, removing shared noise and enabling more accurate detection of context-specific genetic effects.

FastGxC is also **robust to missing data** â€”for example, when certain individuals or genes are missing in some cell types or tissues.

Please read the [BioRxiv](https://www.biorxiv.org/content/10.1101/2021.06.17.448889v2) preprint for more details. 

<!-- Extended data with FastGxC results on GTEx, OneK1K, and CLUES cohorts can be found [here](https://zenodo.org/record/5015123#.YNJ1WpNKjOR) -->


## Please see the [Wiki](https://github.com/BalliuLab/FastGxC/wiki) for a tutorial on how to install and run FastGxC on simulated and real data.



